src/Features/Core/Portable/Wrapping/AbstractSeparatedListCodeComputer.cs

```diff
@@ -0,0 +1,196 @@
+
+            protected async Task AddWrappingGroups(ArrayBuilder<WrappingGroup> result)
+            {
+                result.Add(await GetWrapEveryGroupAsync().ConfigureAwait(false));
+                result.Add(await GetUnwrapGroupAsync().ConfigureAwait(false));
+                result.Add(await GetWrapLongGroupAsync().ConfigureAwait(false));
+            }
+
+            #region unwrap group
+
+            protected abstract Task<WrappingGroup> GetUnwrapGroupAsync();
+
+            protected abstract Task<WrapItemsAction> GetUnwrapAllCodeActionAsync(string parentTitle, WrappingStyle wrappingStyle);
+
+            protected virtual ImmutableArray<Edit> GetUnwrapAllEdits(WrappingStyle wrappingStyle)
+            {
+                var result = ArrayBuilder<Edit>.GetInstance();
+
+                AddTextChangeBetweenOpenAndFirstItem(wrappingStyle, result);
+
+                foreach (var comma in _listItems.GetSeparators())
+                {
+                    result.Add(Edit.DeleteBetween(comma.GetPreviousToken(), comma));
+                    result.Add(Edit.DeleteBetween(comma, comma.GetNextToken()));
+                }
+
+                result.Add(Edit.DeleteBetween(_listItems.Last(), _listSyntax.GetLastToken()));
+                return result.ToImmutableAndFree();
+            }
+
+            #endregion
+
+            #region wrap long line
+
+            protected abstract Task<WrappingGroup> GetWrapLongGroupAsync();
+
+            protected async Task<WrapItemsAction> GetWrapLongLineCodeActionAsync(
+                string parentTitle, WrappingStyle wrappingStyle)
+            {
+                var indentationTrivia = GetIndentationTrivia(wrappingStyle);
+
+                var edits = GetWrapLongLinesEdits(wrappingStyle, indentationTrivia);
+                var title = GetNestedCodeActionTitle(wrappingStyle);
+
+                return await TryCreateCodeActionAsync(edits, parentTitle, title).ConfigureAwait(false);
+            }
+
+            protected abstract ImmutableArray<Edit> GetWrapLongLinesEdits(
+                WrappingStyle wrappingStyle, SyntaxTrivia indentationTrivia);
+
+            #endregion
+
+            #region Wrap every
+            protected abstract Task<WrappingGroup> GetWrapEveryGroupAsync();
+
+            protected async Task<WrapItemsAction> GetWrapEveryNestedCodeActionAsync(
+                string parentTitle, WrappingStyle wrappingStyle)
+            {
+                var indentationTrivia = GetIndentationTrivia(wrappingStyle);
+
+                var edits = GetWrapEachEdits(wrappingStyle, indentationTrivia);
+                var title = GetNestedCodeActionTitle(wrappingStyle);
+
+                return await TryCreateCodeActionAsync(edits, parentTitle, title).ConfigureAwait(false);
+            }
+
+            protected virtual ImmutableArray<Edit> GetWrapEachEdits(
+                WrappingStyle wrappingStyle, SyntaxTrivia indentationTrivia)
```

Comment:
virtual?

Reply:
I made it abstract and implemented the logic in the subclasses.